reactor-netty-http/src/main/java/reactor/netty/http/client/Http2ConnectionLivenessHandler.java

```diff
@@ -0,0 +1,285 @@
+				if (log.isDebugEnabled()) {
+					log.debug("{} channel is currently reading or writing data.", channel);
+				}
+
+				isPingAckPending = false;
+				pingAckDropCount = 0;
+				pingScheduler = invokeNextSchedule();
+				return;
+			}
+
+			if (!isPingAckPending) {
+				if (log.isDebugEnabled()) {
+					log.debug("Attempting to send a ping frame to {} channel.", channel);
+				}
+
+				writePing(ctx);
+				pingScheduler = invokeNextSchedule();
+				return;
+			}
+
+			if (isOutOfTimeRange()) {
+				countPingDrop();
+
+				if (isExceedAckDropThreshold()) {
+					if (log.isInfoEnabled()) {
+						log.info("Closing {} channel due to delayed ping frame response (timeout: {} ns). lastReceivedPingTime: {}, current: {}", channel, pingAckTimeoutNanos, lastReceivedPingTime, System.nanoTime());
+					}
+
+					close(channel);
+					return;
+				}
+
+				if (log.isInfoEnabled()) {
+					log.info("Drop ping ack frame in {} channel. (ping: {})", channel, lastSentPingData);
+				}
+
+				writePing(ctx);
+				pingScheduler = invokeNextSchedule();
+				return;
+			}
+
+			isPingAckPending = false;
+			pingAckDropCount = 0;
+			pingScheduler = invokeNextSchedule();
+		}
+
+		private void writePing(ChannelHandlerContext ctx) {
+			lastSentPingData = ThreadLocalRandom.current().nextLong();
+
+			encoder.frameWriter()
+					.writePing(ctx, false, lastSentPingData, ctx.newPromise())
+					.addListener(pingWriteListener);
+			ctx.flush();
+		}
+
+		private boolean isIoInProgress() {
+			return pingAckTimeoutNanos >= (System.nanoTime() - lastIoTime);
+		}
+
+		private boolean isOutOfTimeRange() {
+			return pingAckTimeoutNanos < Math.abs(lastReceivedPingTime - lastSendingPingTime);
+		}
+
+		private void countPingDrop() {
+			pingAckDropCount++;
+		}
+
+		private boolean isExceedAckDropThreshold() {
+			return pingAckDropCount > pingAckDropThreshold;
+		}
+
+		private ScheduledFuture<?> invokeNextSchedule() {
+			return ctx.executor()
+					.schedule(
+							new PingChecker(ctx),
+							pingScheduleIntervalNanos,
+							NANOSECONDS
+					);
+		}
+
+		private void close(Channel channel) {
+			channel.close()
+					.addListener(future -> {
+						if (future.isSuccess()) {
+							if (log.isDebugEnabled()) {
+								log.debug("{} channel closed an channel", channel);
+							}
+						}
+						else if (log.isDebugEnabled()) {
+							log.debug("{} channel failed to close an channel", channel, future.cause());
+						}
+					});
+		}
+	}
+
+	private class PingWriteListener implements ChannelFutureListener {
+
+		@Override
+		public void operationComplete(ChannelFuture future) throws Exception {
+			if (future.isSuccess()) {
+				if (log.isDebugEnabled()) {
+					log.debug("Wrote PING frame to {} channel.", future.channel());
+				}
+
+				isPingAckPending = true;
+				lastSendingPingTime = System.nanoTime();
+			}
+			else if (log.isDebugEnabled()) {
+				log.debug("Failed to wrote PING frame to {} channel.", future.channel());
+			}
+		}
+	}
```

Comment:
If the new Ping frame is successfully sent, the status is changed to 'ping awaiting', and `lastSendingPingTime` is updated to the current time.